Fundamentals concept (#190)
exercism/awk: concepts/fundamentals/about.md

- split the record into fields using whitespace as the "field separator", so each _word_ is a field.

## Patterns and Actions, continued

The pattern can be omitted, in which case the action is evaluated _for every input record_.

The action can be omitted, in which case the default action is to print the current record.

### Special patterns

There are two special patterns that are often used.

- `BEGIN` will evaluate the associated action before any input is consumed.
- `END` will evaluate the associated action after all input is consumed.

## Field variables

You refer to fields as `$1`, `$2`, `$3`, etc.

The current record is `$0`.

Here are some other frequently used builtin variables.

### Builtin variables that control input parsing

- `RS` is the input record separator. As mentioned, its default value is a newline.
- `FS` is the input field separator.

These variable are often set in the `BEGIN` action block.

### Builtin variables that convey information

These variables are automatically set by AWK during execution of the program.

- `NF` stores the number of fields in the current record.

> - `$NF` is the value of the last field.

Reviewer: `$NF` isn't a variable. `NF` is a variable. `$` is used to access a field. Those are separate concepts.

- `NR` stores the current record number.

>   - when multiple input files are being processed, `NR` will hold the total number of records processed so far.

Reviewer: ```suggestion
  - When multiple input files are being processed, `NR` will hold the total number of records processed so far.
```

> - `FNR` stores the current record number of the current input file
>   - a common AWK idiom is

Reviewer: ```suggestion
- `FNR` stores the current record number of the current input file.
  - Here is common AWK idiom to check if we are processing the first input file.
```

    ```awk
    NR == FNR {action}
    ```
    This means "if the record number equals the record number of the current file", which is only true for records in the _first input file_.

## Truthiness

> AWK was developed at Bell Labs, where C was invented.
> AWK uses the same concept of truthiness as C:
> 
> - the number `0` is false
> - any other number is true

Reviewer: ```suggestion
AWK uses the same concept of truthiness as C.
(AWK was developed at Bell Labs, where C was invented.)

- the number `0` is false.
- any other number is true.
```

Since AWK is a text-processing language, truth values are extended to strings.

> - an empty string `""` is false
> - any non-empty string is true

Reviewer: ```suggestion
- An empty string `""` is false.
- Any non-empty string is true.
```
